### ISSUE TITLE
inspector.has_table throws for temp table in Azure Synapse Dedicated Pool

### ISSUE THREAD

**Issue by jabbera** [sqlalchemy/sqlalchemy#8714]
### Describe the bug

```python
with engine.begin() as conn:
    insp = sqlalchemy_inspect(conn)
    insp.has_table("#blah")
```

throws when used against azure synapse dedicated sql pools. This unfortunately breaks pandas DataFrame.to_sql method.

Microsoft has documentation how how to check for temp tables which I validated works:

https://learn.microsoft.com/en-us/azure/synapse-analytics/sql-data-warehouse/sql-data-warehouse-tables-temporary

```
SELECT OBJECT_ID('tempdb..#blah')
```

### To Reproduce

```python
import logging
import sqlalchemy
from sqlalchemy import inspect as sqlalchemy_inspect

logging.basicConfig(level=logging.DEBUG)

connection_string = "DRIVER={ODBC Driver 18 for SQL Server};Server=<SNIP>;Database=<SNIP>;UID=<SNIP>;PWD=<SNIP>"
url = sqlalchemy.engine.URL.create(
    "mssql+pyodbc",
    query={"odbc_connect": connection_string, "autocommit": "True"},
)

engine = sqlalchemy.engine.create_engine(url)
engine.execution_options(isolation_level="AUTOCOMMIT")

with engine.begin() as conn:
    insp = sqlalchemy_inspect(conn)
    insp.has_table("#blah")
```

### Error

```
do_execute (c:\sandbox\sqltest\.venv\Lib\site-packages\sqlalchemy\engine\default.py:736)
_execute_context (c:\sandbox\sqltest\.venv\Lib\site-packages\sqlalchemy\engine\base.py:1943)
raise_ (c:\sandbox\sqltest\.venv\Lib\site-packages\sqlalchemy\util\compat.py:208)
_handle_dbapi_exception (c:\sandbox\sqltest\.venv\Lib\site-packages\sqlalchemy\engine\base.py:2124)
_execute_context (c:\sandbox\sqltest\.venv\Lib\site-packages\sqlalchemy\engine\base.py:1943)
_execute_clauseelement (c:\sandbox\sqltest\.venv\Lib\site-packages\sqlalchemy\engine\base.py:1572)
_execute_on_connection (c:\sandbox\sqltest\.venv\Lib\site-packages\sqlalchemy\sql\elements.py:333)
execute (c:\sandbox\sqltest\.venv\Lib\site-packages\sqlalchemy\engine\base.py:1380)
has_table (c:\sandbox\sqltest\.venv\Lib\site-packages\sqlalchemy\dialects\mssql\base.py:2947)
_switch_db (c:\sandbox\sqltest\.venv\Lib\site-packages\sqlalchemy\dialects\mssql\base.py:2600)
wrap (c:\sandbox\sqltest\.venv\Lib\site-packages\sqlalchemy\dialects\mssql\base.py:2576)
has_table (c:\sandbox\sqltest\.venv\Lib\site-packages\sqlalchemy\engine\reflection.py:283)
<module> (c:\sandbox\sqltest\test_new.py:37)
_run_code (c:\Python\310\Lib\runpy.py:86)
_run_module_as_main (Current frame) (c:\Python\310\Lib\runpy.py:196)
``` 

sqlalchemy.exc.ProgrammingError: (pyodbc.ProgrammingError) ('42S02', "[42S02] [Microsoft][ODBC Driver 18 for SQL Server][SQL Server]Invalid object name 'tempdb.pdw.INFORMATION_SCHEMA_COLUMNS'. (208) (SQLExecDirectW)") 
[SQL: SELECT tempdb.[INFORMATION_SCHEMA].[COLUMNS].[TABLE_NAME]
FROM tempdb.[INFORMATION_SCHEMA].[COLUMNS]
WHERE tempdb.[INFORMATION_SCHEMA].[COLUMNS].[TABLE_NAME] LIKE CAST(? AS NVARCHAR(max))]
[parameters: ('#blah[_][_][_]%',)]

### Versions

- OS: Windows 10
- Python: python 3.10
- SQLAlchemy: 2.0.0b2
- Database: Azure Synapse Dedicated Pool
- DBAPI (eg: psycopg, cx_oracle, mysqlclient): pyodbc/msft

**Comment by zzzeek** [comment]
hey there -

im not sure this is something we could change outside of a major release, unless the same method works for all SQL Server databases.  but also, has_table() is just part of a much bigger reflection picture.  if azure does not provide any function for retreiving full column info for temp tables then we might have to leave temp table reflection not implemented for now.

cc @gordthompson

**Comment by jabbera** [comment]
I actually think this is broken in general for sql server. The query used by default show temp tables created by anyone. IE I could create a temp table called #blah and that's only valid for my session. The query being used would make it to everyone that the table exists.

**Comment by jabbera** [comment]
I dug through the code and I see there is a filter on object_id, but this implementation of has_table seems very complicated for no reason:

https://github.com/sqlalchemy/sqlalchemy/blob/eae9d1420bbfde4dbd835b654e80653cd5ac2155/lib/sqlalchemy/dialects/mssql/base.py#L3229-L3239

The first query that's hanging me up is totally unnecessary. The subsequent query is the only part that's needed.

https://github.com/sqlalchemy/sqlalchemy/blob/eae9d1420bbfde4dbd835b654e80653cd5ac2155/lib/sqlalchemy/dialects/mssql/base.py#L3243-L3248

**Comment by gordthompson** [comment]
Hi @jabbera . Thanks for this! Have you tested your patch against a real Azure instance (we don't have one available for testing)? If so, then we can push your PR to Gerrit for CI testing.

**Comment by zzzeek** [comment]
@gordthompson the patch seems to simplify things in any case for plain SQL Server also, so provided it still works (Seems to?) we would be able to merge this for main / 1.4 I think.   do you see any risk ?

**Comment by gordthompson** [comment]
> we would be able to merge this for main / 1.4 I think. do you see any risk ?

We could at least try to cherry-pick it and hope it doesn't cause too many conflicts with the multi-reflect changes.

**Comment by CaselIT** [comment]
I no expert in mssql or azire things, but won't `SELECT object_id(:table_name)` transform return if there is any object, not just tables and views?

**Comment by jabbera** [comment]
> I no expert in mssql or azire things, but won't `SELECT object_id(:table_name)` transform return non null if there is any object, not just tables and views?

Yes but in this case the filter removed requires the object be prefixed with # which means it must be a temp table.

**Comment by jabbera** [comment]
> Hi @jabbera . Thanks for this! Have you tested your patch against a real Azure instance (we don't have one available for testing)? If so, then we can push your PR to Gerrit for CI testing.

Yes. I tested it. Additionally you can see by looking at the patch that the code removed adds nothing useful. It basically says: if a temp table exists in a metadata table that lets me look up all temp tables from all sessions, go and check if one exists just for my session (using object_id). There is zero reason to not just use object_id first. There is probably a race condition in the existing solution that's removed in my implementation.

**Comment by CaselIT** [comment]
> > I no expert in mssql or azire things, but won't `SELECT object_id(:table_name)` transform return non null if there is any object, not just tables and views?
> 
> Yes but in this case the filter removed requires the object be prefixed with #. There is an if block enclosing the code which means my change only applies to temp tables. It's a pretty narrow fix. Additionally I didn't add the object_id code. It was already there for temp tables. I just changed it to be unconditionally called as it's a more accurate way to determine if a temp table exists then the previous implementation.

Ok, if only temp tables can start with `#` and be placed in `tempdb` this looks ok to me

**Comment by jabbera** [comment]
> Ok, if only temp tables can start with `#` and be placed in `tempdb` this looks ok to me

From MSFT:

An identifier that starts with double number signs (##) denotes a global temporary object. Although the number sign or double number sign characters can be used to begin the names of other types of objects, we do not recommend this practice.

I'll update my fix to filter for just tables.

**Comment by CaselIT** [comment]
Note that also views and marerialized views are returned by has table

**Comment by jabbera** [comment]
@CaselIT I just pushed a change that restricts the result to user tables. See documentation here:

https://learn.microsoft.com/en-us/sql/relational-databases/system-catalog-views/sys-objects-transact-sql?view=sql-server-ver16

Sql server doesn't support temporary views. I saw the comment in the code and just double checked:

Msg 4103, Level 15, State 1, Line 1
"#A": Temporary views are not allowed.

Completion time: 2022-10-26T08:13:20.7112753-04:00
